prov/verbs/src/verbs_cq.c

```diff
@@ -36,39 +36,46 @@
 
 #include "fi_verbs.h"
 
-static uint64_t fi_ibv_comp_flags(struct ibv_wc *wc)
+static inline uint64_t fi_ibv_handle_imm_data(struct ibv_wc *wc, uint64_t *data)
 {
-	uint64_t flags = 0;
-
-	if (wc->wc_flags & IBV_WC_WITH_IMM)
-		flags |= FI_REMOTE_CQ_DATA;
+	if (wc->wc_flags & IBV_WC_WITH_IMM) {
+		if (data)
+			*data = ntohl(wc->imm_data);
+		return FI_REMOTE_CQ_DATA;
+	} else {
+		if (data)
+			*data = 0;
+		return 0;
+	}
+}
 
+static inline uint64_t fi_ibv_comp_flags(struct ibv_wc *wc, size_t *len)
```

Comment:
Please rename to something like fi_ibv_handle_wc(wc, &flags, &len, &data) and merge this function with fi_ibv_handle_imm_data() above.  This should provide another level of optimization, since immediate data only applies to IBV_WC_RECV_RDMA_WITH_IMM or IBV_WC_RECV opcodes.  Since 3 different fields are being set -- flags, data, and len -- I would just pass all into the function.

Reply:
Thanks! I like your suggestion. Done